pyomo/contrib/alternative_solutions/lp_enum_solnpool.py

```diff
@@ -0,0 +1,235 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright (c) 2008-2022
+#  National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+try:
+    from gurobipy import GRB
+
+    gurobi_available = True
+except:
+    gurobi_available = False
+
+import pyomo.environ as pe
+from pyomo.contrib.alternative_solutions import aos_utils, shifted_lp, solution
+from pyomo.contrib import appsi
+
+
+class NoGoodCutGenerator:
+    def __init__(
+        self,
+        model,
+        variable_groups,
+        zero_threshold,
+        orig_model,
+        all_variables,
+        orig_objective,
+        num_solutions,
+    ):
+        self.model = model
+        self.zero_threshold = zero_threshold
+        self.variable_groups = variable_groups
+        self.variables = aos_utils.get_model_variables(model)
+        self.orig_model = orig_model
+        self.all_variables = all_variables
+        self.orig_objective = orig_objective
+        self.solutions = []
+        self.num_solutions = num_solutions
+
+    def cut_generator_callback(self, cb_m, cb_opt, cb_where):
+        if cb_where == GRB.Callback.MIPSOL:
+            cb_opt.cbGetSolution(vars=self.variables)
+            print("***FOUND SOLUTION***")
+
+            for var, index in self.model.var_map.items():
+                var.set_value(var.lb + self.model.var_lower[index].value)
+            sol = solution.Solution(
+                self.orig_model, self.all_variables, objective=self.orig_objective
+            )
+            self.solutions.append(sol)
+
+            if len(self.solutions) >= self.num_solutions:
+                # TODO: (nicely) terminate the solve
+                return
+
+            num_non_zero = 0
+            non_zero_basic_expr = 1
+            for idx in range(len(self.variable_groups)):
+                continuous_var, binary_var = self.variable_groups[idx]
+                for var in continuous_var:
+                    if continuous_var[var].value > self.zero_threshold:
+                        num_non_zero += 1
+                        non_zero_basic_expr += binary_var[var]
+            # TODO: JLG - If we want to add the mixed binary case, I think we
+            # need to do it here. Essentially we would want to continue to
+            # build up the num_non_zero as follows
+            # for binary in binary_vars:
+            # if binary.value > 0.5:
+            # num_non_zero += 1 - binary
+            # else:
+            # num_non_zero += binary
```

Comment:
Needed?

Reply:
Same as earlier.  These comments refer to features we haven't flushed out yet.